awx db partitioning

jladdjr opened the pull request:
closing this PR, opened [new PR](https://github.com/ansible/awx/pull/9839) using branch under ansible org to make it easier for @chrismeyersfsu and I to work on together

Commit:
first draft of db partitioning

Commit:
Rename / remove old main_jobevent table

Commit:
Register new column created by SQL

* .. using FakeAddField model type
* .. without doing this, Django won't know
     about the field we created using raw SQL

Commit:
By default, save job_created as null in db

* Want to avoid saving date as empty string
* Should default to null so that it's obvious the field is empty

Commit:
When copying main_jobevent, include all table metadata

* copy the table just like we do in the bigint migration
* without this we lose sequences and very likely other things
  as well
* we want the new table to be identical to the old table,
  so 'including all' makes sense

Commit:
Create fake partition

* Just to get things working
* Will implement dynamic creation of partitions later

Commit:
Drop primary key index before creating partition table

* Partition tables require unique contstraints to include the
  partition key (uniqueness can only be enforced _inside_
  of a given partition table)

Commit:
When saving JobEvents, include job_created

* this is the partition key
* .. used to determine which partition job event rows are sent to

Commit:
Convert job_created to string for serialization

Commit:
Only save job_created field on JobEvent for the time being

* Once other job event tables are migrated, remove this

Commit:
Remove temporary catch-all partition

Commit:
Dynamically create initial partitions

* First partition holds all events up to this very moment
* And second partition starts where first left off and runs
  .. through rest of current hour
* From there, dynamically generated partitions will cover
  one hour at a time

jladdjr commented:
cc @Ladas - for some reason the 'request review' feature isn't finding your name. cc-ing you for awareness

Commit:
first partition is _not_ a special case after all

Commit:
create helper method to create partitions

* create_partition()

Ladas commented:
> I guess that's true - if someone did run a job immediately after the migration, the new migration that was created would overlap with the old partition, but any events created would just go in the new partition.
> 
> okay, cool, so this isn't needed

Seem like you figured it out, so just a note: the partitions can't overlap. Postgre should fail if you'd try to create overlapping partition. Same as it will fail if no partition is available for the timestamp.

Commit:
Give new primary key constraint unique name, create first live partition

Commit:
Create partition only if it doesn't exist

Commit:
auto-create partition

Commit:
fix import

Commit:
add option to create partitions that span a single minute

* turned on by default currently for testing

Commit:
Move partition start date to 2000

Commit:
Only create partitions for regular jobs

Commit:
get job events based on job creation date

ryanpetrello commented on awx/main/migrations/0124_event_partitions.py:
```diff
@@ -0,0 +1,112 @@
+
+    # Only partitioning main_jobevent on first pass
+    #
+    #for tblname in (
```
@jladdjr before you continue onward with with problems related to analytics, job event cleanup, and large-scale migrations, it might be good to pause and update this PR with some changes that introduce partitions for *all* event types - looking at the layout of this code and the comments here, I think you're pretty close 😄, and having that is probably a first pass of what QE and perfscale folks need to actually stand up and test this from a functional/correctness perspective.

ryanpetrello commented on awx/main/utils/common.py:
```diff
@@ -1079,3 +1082,47 @@ def truncate_stdout(stdout, size):
+            partition_label = start.strftime('%Y%m%d_%H')
+
+    with connection.cursor() as cursor:
+        # Only partitioning main_jobevent on first pass
```
Similar to the "make it work for all use cases" comment above, you could probably lose this loop, and just pass `tblname` _directly into_ this function so that the proper parent table is operated on depending on the context from which this function is called (either in the upgrade-time migration, or via the task manage as new jobs are scheduled).

ryanpetrello commented on awx/main/scheduler/task_manager.py:
```diff
@@ -255,6 +256,11 @@ def start_task(self, task, rampart_group, dependent_tasks=None, instance=None):
+        # TODO: Currently, only creating partition for jobs. Drop contiditional once
+        # there are partitions for all job event types
+        if task_actual['type'] == 'job':
+            create_partition()
```
I *do* think this is a reasonable place to put this function call, because it's within the scope of the global task manager advisory lock (so we don't have to worry about competing jobs creating a partition twice in close succession), and it's *right before* we anticipate we'll need the partition.

This doesn't need to happen now, but we should think about what it would mean if this function called *threw some exception*.  How would an uncaught exception behave in this portion of the task manager?  Would we see an error?  Would the task manager exit gracefully and "try again" in 30s?  Is that okay?

cc @chrismeyersfsu @Ladas 

This is a complicated and sensitive part of the overall system, so we should start pondering on it early 😄

ryanpetrello replied:
It might even make more sense to put it at the top of the "post commit" function below where we actually *are positive* we're about to run an `.apply_async()` call:

https://github.com/ansible/awx/pull/9341/files#diff-a37220424979b0075fa6e25bf8c309d671f30641e64830c27cd046272c73a703R333

That way if the job fails `.pre_start()` (https://github.com/ansible/awx/pull/9341/files#diff-a37220424979b0075fa6e25bf8c309d671f30641e64830c27cd046272c73a703R268) or any of the other code directly below _this_ section of the code, and we *don't* actually end up passing a message to the dispatcher to perform work, we won't have needlessly created a partition.

Doing it that way will also ensure that we don't create partitions for `WorkflowJob`s, which don't have events.

ryanpetrello replied:
Sounds like we're also ready to call this function for _all_ event types, not just job events?

jladdjr replied:
yep, got pulled off before I was able to make this change 👍

Commit:
Misc doc clean-up

Commit:
Partition *all* job event tables

ryanpetrello commented on awx/main/migrations/0124_event_partitions.py:
```diff
@@ -0,0 +1,107 @@
+
+            # .. as well as initial partition containing all existing events
+            epoch = datetime(2000, 1, 1, 0, 0)
+            create_partition(epoch, current_time, 'old_events')
```
`create_partition(tblname, ...)` ?

jladdjr replied:
yep, I got pulled off before I got everything updated / pushed / tested

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1388,6 +1393,10 @@ def run(self, pk, **kwargs):
         if self.instance.spawned_by_workflow:
             self.parent_workflow_job_id = self.instance.get_workflow_job().id
 
+        # TODO: can we count on instance always having created?
```
Yes.  If we get to this point where we're creating an actual event, the related job already exists in the database, and definitely has a `created` value.

ryanpetrello replied:
Also, this column is not nullable:

```sql
                                             Table "public.main_unifiedjob"
         Column          |           Type           | Collation | Nullable |                   Default
-------------------------+--------------------------+-----------+----------+---------------------------------------------
 id                      | integer                  |           | not null | nextval('main_unifiedjob_id_seq'::regclass)
 polymorphic_ctype_id    | integer                  |           |          |
 created                 | timestamp with time zone |           | not null |
```

jladdjr replied:
The db as the single source of truth on existence of data - that's cool / makes sense 👍 

Also, makes sense that the job _better_ have a created value. It was just that as I was going through keying off of things I wondered about taking for granted the existence of fields. Good to know ^

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1246,6 +1247,10 @@ def event_handler(self, event_data):
             event_data['workflow_job_id'] = self.parent_workflow_job_id
+        # Do we have to check if the field exists? if it doesn't
+        # how will be eventually store the event in the db?
+        if self.job_created:
```
Yep, I'd say we don't want this if statement, because if this value isn't set, the event won't be saved (because it's what the partitioning is based on).

If you get to this part of the codebase and `self.job_created` isn't set, there's a bug (so let's expose it).

jladdjr replied:
makes sense - dropped comment / conditional

ryanpetrello commented on awx/api/views/__init__.py:
```diff
@@ -3826,7 +3826,7 @@ class JobJobEventsList(BaseJobEventsList):
         job = self.get_parent_object()
         self.check_parent_access(job)
-        qs = job.job_events.select_related('host').order_by('start_line')
+        qs = job.job_events.select_related('host').filter(job_created=job.created).order_by('start_line')
```
We should copy this approach for all of the event list endpoints e.g.,

`/api/v2/project_updates/N/events/`
`/api/v2/inventory_updates/N/events/`
`/api/v2/system_job_updates/N/events/`
`/api/v2/ad_hoc_commands/N/events/`

jladdjr replied:
If we wanted to put this method _everywhere_ seemed like maybe we wanted it in BaseJobEventsList
.. but not all of the other classes are children of that so 🤷 
.. but maybe there's another reason why we actually want this copy/pasted in the other classes?

jladdjr replied:
oh, and it looks like in some cases the qs is formed a bit differently (`HostJobEventsList(BaseJobEventsList)`)

jladdjr replied:
pushed https://github.com/ansible/awx/pull/9341/commits/1f746733cfeea99a98208a4fda4703e9109c7f3b - I'll need to test this out

Commit:
create_partition needs tblname

Commit:
Drop todo / answered question

Commit:
Drop todo / question / conditional

* can safely assume job_created is set
* .. and if it isn't, we want to expose that bug

Commit:
Update querysets for remaining job event views

- search should use job creation date
  so that only the table partition with the relevant job events
  is searched

ryanpetrello commented on awx/main/models/events.py:
```diff
@@ -426,13 +426,32 @@ def create_from_data(cls, **kwargs):
             kwargs.pop('created', None)
 
+        # same as above, for job_created
+        # TODO: if this approach, identical to above, works, can convert to for loop
```
I think doing this explicitly in this way is fine without a loop

ryanpetrello commented on awx/main/utils/common.py:
```diff
@@ -1079,3 +1082,38 @@ def truncate_stdout(stdout, size):
+    - end defaults to end of current hour
+    - partition_label defaults to YYYYMMDD_HH
+
+    - minutely will create partitions that span _a single minute_ for testing purposes
```
This is neat and helpful for testing and prototyping; let's make a mental note to remove the per-minute support prior to merge, however (having this accidentally called at some point in the future would be bad, because postgres isn't going to like having this many partitions over time.  Also, we're likely to have other logic (event cleanup) that assumes partitions are "per hour").

Commit:
fix initial partition creation

* call create_partition correctly (include tblname)
* reference 'awx_epoch'

Commit:
Fix migration for _all_ job event tables

- each job event table has a different name for the
  fk referencing the unified job id
- create a mapping so that we get the col name correct
  for each table

Commit:
Update fields for job event models to match raw sql operations

* raw sql commands were in migration to partition table
* .. just needed to add FakeAddField entries for the new
     job_created field added to each job event model
* .. and also needed to actually list the new field on the model classes

Commit:
Continue updating job event views

Commit:
include job_created field in all job event types

ryanpetrello commented on awx/main/migrations/0124_event_partitions.py:
```diff
@@ -0,0 +1,133 @@
+            # TODO: do more generic search for pkey constraints
+            # instead of hardcoding this one that applies to main_jobevent
+            cursor.execute(
+                f'ALTER TABLE {tblname}_old DROP CONSTRAINT {tblname}_pkey1'
```
Reading pg docs, this is a little unclear to me.  Does the *primary key* have to include the partition column, or just **any** indexed constraint?  There's almost certainly going to be situations where we want to search for a specific event based on ID only e.g., `/api/v2/jobs/N/job_events/Y/` and it's going to be super slow if this column isn't the key.

e.g.,

Does this still work if we **keep** `id` as the primary key, and *instead* do:

```sql
ALTER TABLE ONLY {tblname} ADD CONSTRAINT {tblname}_unique_id_job_created UNIQUE (id, job_created);
```

If this works w/ partitions, I'd prefer it.

jladdjr replied:
> Does this still work if we keep id as the primary key

No, this is not an option; postgres refuses to let you claim a field as being unique _unless_ you use a composite primary key and include the partition field (job_created)

jladdjr replied:
okay, found my comment where I posted my findings on this:
https://github.com/ansible/awx/issues/9259#issuecomment-776325677

jladdjr replied:
What I _think_ pg will try to do using the tables exactly the way they are today is:
* run a parallel search of each of the partitions associated with the parent job event table (literally spinning up different workers to handle each table partition)
* for each partition, use _that table's_ index for id / job_created

I tried running a query with `explain analyze` just now, but I hardly have any events in my db. It's showing a sequential scan, but my sense is that's probably because the postgres planner knew we hardly had any events and didn't bother with the index.

cc @Ladas for fact-checking because this is a pretty wild guess on my part

Ladas replied:
@ryanpetrello @jladdjr so for query like `/api/v2/jobs/N/job_events/Y/` you will still need to use the partition key `job_created`, otherwise PG will scan all partitions and that will always be slow.

So the rule of thumb is using the partitioning key in all queries, when using partitions. (Well there are e.g. cases where you'd want to scan all the data in parallel, but postgres does pretty good parallel seq scan even without partitions.)

For this specific usecase PG requires unique constraint/primary key to include the partitioning key. And it should use the index on id with `(id, job_created)`

-------

@jladdjr yeah, the index/seq scan is dependent on what cost will PG compute. Which in simplified terms takes into account what percentage of the table PG thinks you'll be looking at, based on statistics it keeps. (very simply said, index should be used when querying ~5% of the table). 

Not sure if you're using default costs for PG. The most important is usually `random_page_cost`, which is by default 4.0, which is good for magnetic disks (index access is expensive). I set it 1.1 for ssd. Do you recommend PG config to users in docs? Or this is up to users?

Commit:
Push changes before rebasing

ryanpetrello committed:
move the partition data migration to be a post-upgrade async process

this copies the approach we took with the bigint migration

ryanpetrello committed:
some more events view cleanup

jladdjr commented:
rebased.. this PR has now joined 2021

ryanpetrello committed:
correct filter events on the migration event horizon

events that existed *prior* to the partition migration will have
`job_created=1970-01-01` auto-applied at migration time; as such,
queries for these events e.g., /api/v2/job/N/job_events/
use 1970-01-01 in related event searche

events created *after* the partition migration (net-new playbook runs
will have `job_created` values that *exactly match* the related
`UnifiedJob.created` field.

ryanpetrello commented on awx/main/migrations/0124_event_partitions.py:
```diff
@@ -0,0 +1,117 @@
+            )
+
+            # let's go ahead and add and subtract a few indexes while we're here
+            cursor.execute(f'CREATE INDEX {tblname}_modified_idx ON {tblname} (modified);')
```
cc @Ladas

Ladas replied:
@ryanpetrello I'm fairly sure this will work for analytics. We'll probably need to limit the query with the partition key (created), so e.g. we scan max 24 partitions/h back. But I think the index should be still used in this case

ryanpetrello replied:
So one question @Ladas is how we're *actually* going to make use of this in a forward-facing way that's *also* backwards-compatible given https://github.com/ansible/awx/pull/9358/files#r584917954

ryanpetrello commented on awx/main/utils/common.py:
```diff
@@ -161,6 +165,14 @@ def memoize_delete(function_name):
+def get_event_partition_epoch():
+    from django.db.migrations.recorder import MigrationRecorder
+    return MigrationRecorder.Migration.objects.filter(
+        app='main', name='0124_event_partitions'
```
@jladdjr if we end up reordering this migration (I'll bet exec envs beat us), we'll have to correct this label.

ryanpetrello committed:
remove the global `/api/v2/job_events/` endpoint

Commit:
Rev migration number

Commit:
Update reference to partition migration

jladdjr commented:
rebased and ran through basic sanity check, confirmed things are still looking good

Commit:
Do not cascade delete job_events

* want to drop job_event _partitions_
  .. instead of having the job events associated with a job
     automatically get cleaned up for us

kdelee commented on awx/main/tasks.py:
```diff
@@ -740,22 +746,16 @@ def update_host_smart_inventory_memberships():
         def _remaining():
             try:
-                cursor.execute(f'SELECT MAX(id) FROM _old_{tblname};')
+                cursor.execute(f'SELECT MAX(id) FROM _unpartitioned_{tblname};')
```
what if we just look if there is at least 1 record
```
select id from _unpartitioned_main_jobevent limit 1;
```

Ladas replied:
@jladdjr  each of the MAX(id), INSERT and DELETE queries is causing sequential scan now, so I think the index on id is needed, if you want to use the current way 

(although there should be a way to move the data with DB procedure without index, in one sequential pass of the table, using a DB cursor :-), but it's better to stay with working algorithm as @ryanpetrello  says :-) )

------

I'd say `copy original schema to TMP table` -> `change TMP table schema(remove the primary key)` -> `create partition parent from TMP table schema`  is probably the simplest (one line change, or 2 lines, deleting also the TMP table :slightly_smiling_face: )

So keeping the primary key on the original events table and reusing the algorithm sounds the simplest :-)

jladdjr replied:
> I'd say copy original schema to TMP table -> change TMP table schema(remove the primary key) -> create partition parent from TMP table schema is probably the simplest (one line change, or 2 lines, deleting also the TMP table 🙂 )

makes sense - I'm writing this up now and will kick the tires a bit. if things look good we can try to patch things in the scale testing context

jladdjr replied:
https://github.com/jladdjr/awx/commit/91b08176884690ef6f3ec69a3cfada884f597571 creates a tmp table that we modify / use as a basis for the partitioned table's schema

jladdjr replied:
I made a small tweak to that ^, but after that was able to confirm that the migration went through and that we still had an index on the old jobevents table

jladdjr replied:
Going to try to run through a migration w/ a larger set of data and see if we hit any performance issues at a bigger scale (before we go back and try this at true scale)

Ladas commented on awx/main/models/events.py:
```diff
@@ -426,13 +426,28 @@ def create_from_data(cls, **kwargs):
             setattr(event, 'workflow_job_id', workflow_job_id)
+        # shouldn't job_created _always_ be present?
+        # if it's not, how could we save the event to the db?
+        job_created = kwargs.pop('job_created', None)
```
I'd say that inserting None will fail, since it won't match any partition

ryanpetrello replied:
Probably better to just do `kwargs['job_created']` here - if we were ever missing this key at this point in the code, it's a bug and we want to know.

Commit:
create tmp schema to reference when creating partitioned table

* if we use the actual old job events table
  and make tweaks to its schema
  namely, dropping the pkey constraint,
  then when we go to migrate the old job events
  we will be forcing postgres to do a sequential scan
  on the old table, which effectively causes the migration to hang

Commit:
correct constraint name

jladdjr commented:
rebased.. and bumped migration

Commit:
bump partition migration

Commit:
lint fixes

Commit:
update tests to reflect new job_event schema

Commit:
bump partition migration to 0132

jladdjr commented:
rebasing again - another migration dropped in devel

jladdjr commented:
builds are now passing..

Commit:
Enable partition pruning when fetching job's events

ryanpetrello committed:
upgrade to partitions without a costly bulk data migration

keep pre-upgrade events in an old table (instead of a partition)

- instead of creating a default partition, keep all events in special
"unpartitioned" tables
- track these tables via distinct proxy=true models
- when generating the queryset for a UnifiedJob's events, look at the
  creation date of the job; if it's before the date of the migration,
  query on the old unpartitioned table, otherwise use the more modern table
  that provides auto-partitioning

ryanpetrello committed:
remove code that leaves behind old bigint tables on fresh installs

we don't need this code at all anymore - the bigint migration is long
gone, and anybody upgrading to this version of AWX has already
migrated their data

Commit:
short-circuit event_processing_finished for wf jobs

* wf jobs are a wrapper for other jobs
* they do not process their own job events

Commit:
Minor rebase fallout

* fix import
* remove dropped url

Commit:
bump partition migration to 135

Commit:
black formatting

Commit:
update mocks to reflect new migration

* instead of mocking `created_or_epoch` (which no longer exists)
* .. mock `Unified_Job.has_unpartitioned_events`

Commit:
test updates

* when tests create a UnifiedJob and JobEvent,
  the two need to have the same value for job creation time
* some view validation was skipped due to `model` being
  a property in some cases now

Commit:
Ensure `View.model` exists before checking type

Commit:
Move created kwargs to right place

Commit:
add migrations for Unpartitioned{Job}Event proxy models

Commit:
mock has_unpartitioned_events in collection tests

.. just like we do with the main awx tests

Commit:
set event horizon to -1 for empty tables

Commit:
is_partitioned should default to true when pk not set

Commit:
bump db partition migration

jladdjr commented:
closing this PR, opened [new PR](https://github.com/ansible/awx/pull/9839) using branch under ansible org to make it easier for @chrismeyersfsu and I to work on together
